feat: Implement canvas drawing system with role-based permissions

rhino-ty opened the pull request:
## 📂 작업 내용

closes #10

- [x] 캔버스 드로잉 시스템 구현
- [x] 게임 역할별 권한 관리 시스템 구현
- [x] 픽셀 기반 잉크 제한 시스템 추가
- [x] 반응형 디자인 적용

## 💡 자세한 설명

### 1. 캔버스 컴포넌트 구조
- `class-variance-authority`를 활용한 일관된 스타일링 시스템 적용
- 모바일 퍼스트로 설계된 반응형 레이아웃
- 4:3 비율 유지를 위한 aspect-ratio 적용
- 최소 너비 280px 보장
- Canvas UI 컴포넌트와 `GameCanvas` 컴포넌트를 분리하여 관심사 분리
- 커스텀 훅 `useDrawing`을 통해 드로잉 로직 모듈화


### 2. 드로잉 기능(임시)

> [!NOTE]
> 드로잉 기능인 `useDrawing.tsx` 등은 다솔님과 합작해 다시 할 예정입니다!

- 색상 선택기 구현
  - 5가지 기본 색상 지원
  - 선택 상태 표시 (ring 효과)
  - hover/focus/active 상태 애니메이션
- 그리기 도구 컨트롤 구현
  - 브러시 사이즈 슬라이더
  - undo/redo 버튼
  - 펜/채우기 모드 전환
  - 잉크 게이지 표시 : 후술

### 3. 역할 기반 권한 관리

> [!NOTE]
> 추후 Redis + Nest.JS로 게임을 만들 때, 다시 `role` 등 변수명을 다시 작성할 예정입니다.

- TypeScript 타입 시스템을 활용한 엄격한 역할 관리 구현
- 그림꾼/방해꾼/구경꾼 역할별 권한 구분
- 타입 가드를 통한 안전한 타입 체크

### 4. 잉크 시스템
- SVG 기반의 시각적 잉크 게이지 구현
- 픽셀 단위의 정확한 사용량 추적
- 성능 최적화를 위한 useMemo 활용

### 5. 사용성 개선
- ARIA 레이블 적용으로 접근성 보장
- 반응형 툴바 위치 조정
- 키보드 내비게이션 지원
- 상태 변경 시 적절한 피드백 제공: hover 등

## 📗 참고 자료 & 구현 결과

### Desktop

<img width="1552" alt="image" src="https://github.com/user-attachments/assets/982d1cb7-ad08-4469-8ebe-57382bb44b78">

### Mobile

<img width="612" alt="image" src="https://github.com/user-attachments/assets/8aec9e48-6d7d-4b75-9b0f-5ca5452aeb6f">

## 📢 리뷰 요구 사항

- 역할 기반 권한 관리 로직의 타입 안정성 검토 필요

## 🚩 후속 작업
- 추가 색상 팔레트 지원
  - 나중에 자유롭게 색을 선택하는 팔레트가 Floating UI로 해도 괜찮을 듯?
- WebRTC 연동을 위한 인터페이스 준비

## ✅ 셀프 체크리스트
- [x] PR 제목을 형식에 맞게 작성했나요?
- [x] 브랜치 전략에 맞는 브랜치에 PR을 올리고 있나요?
- [x] 이슈는 close 했나요?
- [x] Reviewers, Labels를 등록했나요?
- [x] 작업 도중 문서 수정이 필요한 경우 잘 수정했나요?
- [x] 테스트는 잘 통과했나요?
- [x] 불필요한 코드는 제거했나요?

rhino-ty committed:
feat: Add redo, pen, bucket icon to assets

- To use undo, redo feature
- To use paint tool icons and mode switching

rhino-ty committed:
feat: Add InkGauge component for drawing pixel limit visualization

- Implemented InkGauge component with SVG-based gauge display
- Optimized re-renders using useMemo for ratio calculation
- Included responsive sizing with Tailwind classes
- Added proper ARIA labels for accessibility

rhino-ty committed:
feat: Add Canvas UI component with drawing toolbar and controls

- Created responsive Canvas component with drawing tools interface
- Implemented cva-based variants for container, toolbar, and buttons
- Added color picker with selectable color buttons
- Integrated pen/fill mode toggle with visual feedback
- Included brush size slider for pen mode
- Added undo/redo controls with disabled states
- Incorporated InkGauge component for pixel limit tracking
- Ensured full accessibility with ARIA labels and states
- Added responsive toolbar positioning (bottom/floating)

rhino-ty committed:
feat: Add useDrawing hook for canvas drawing functionality

- Implemented drawing state management with pixel tracking
- Added drawing modes (pen and flood fill) with history support
- Created pixel-based ink limit system with usage tracking
- Added undo/redo functionality using ImageData history
- Implemented basic and flood fill drawing algorithms
- Optimized with useCallback for performance
- Added brush size and color management
- Included canvas initialization and cleanup
- Added touch support for mobile devices

This is an AI-generated example to demonstrate canvas drawing capabilities

rhino-ty committed:
feat: Add example canvas components with user roles like game

- Created type definitions for user statuses and roles (UserStatus, UserRole)
- Added GameCanvas component with role-based drawing permissions
- Implemented type guards for drawable roles
- Added mouse and touch event handlers for drawing
- Set up predefined color palette system
- Included TypeScript utility types for role management
- Connected useDrawing hook with canvas interactions
- Added pixel limit tracking for drawing constraints

Example implementation demonstrates

rhino-ty committed:
Merge branch 'develop' into feature/#10

choiseona commented on client/src/components/canvas/CanvasUI.tsx:
```diff
@@ -0,0 +1,241 @@
+            >
+              <div className="flex gap-1.5">
+                {colors.map((color, index) => (
+                  <Button
+                    key={index}
+                    className={cn(colorButtonVariants({ isSelected: color.isSelected }))}
```
컬러 개수가 고정이라면 key로 index를 부여해도 상관없겠네용

rhino-ty replied:
맞습니당 동적이지 않아서 그냥 index 썼어요

choiseona commented on client/src/components/canvas/CanvasUI.tsx:
```diff
@@ -0,0 +1,241 @@
+                    style={{ backgroundColor: color.backgroundColor }}
+                    size="icon"
+                    onClick={color.onClick}
+                    aria-label={`${color.color} 색상 선택`}
+                    aria-pressed={color.isSelected}
```
웹 접근성 좋습니다 👍 저도 리팩토링 할 때 웹 접근성 학습하면서 적용해봐야겠어요!

choiseona commented on client/src/components/canvas/GameCanvasExample.tsx:
```diff
@@ -0,0 +1,135 @@
+      });
+    },
+    [draw],
+  );
```
공통되는 로직은 함수로 분리하는 것도 괜찮을 것 같습니다!

rhino-ty replied:
UI - 비즈니스 로직 - 상호작용 이런 식으로 나눌 거라 나중에 Canvas UI랑 로직이 안정화되면 그 때 리팩토링 하도록 하죠!

choiseona commented on client/src/hooks/useDrawing.ts:
```diff
@@ -0,0 +1,256 @@
+
+  // 드로잉 상태 저장
+  const saveDrawingState = useCallback(() => {
+    const canvas = canvasRef.current;
+    const ctx = canvas?.getContext('2d');
+    if (!ctx || !canvas) return;
```
이 부분을 반복적으로 사용하고 있는데, 따로 함수로 빼는 건 어떨까요??

rhino-ty replied:
수정 완

rhino-ty committed:
refactor: Refactor event type imports for consistency and readability

choiseona commented on client/src/components/canvas/CanvasUI.tsx:
```diff
@@ -0,0 +1,241 @@
+      colors = [],
+      onUndo,
+      onRedo,
+      onBrushSize,
```
onBrushSize 말고 onChangeBrushSize 어떨까요?

rhino-ty replied:
수정 완.

choiseona reviewed (commented):
고생하셨습니다!

rhino-ty committed:
refactor: Change onBrushSize function name for convention

- Change to setBrushSize

rhino-ty committed:
Merge branch 'develop' into feature/#10

rhino-ty committed:
refactor: Extract canvas context check logic

- Add `getCanvasContext` utility function to centralize canvas context validation
- Add warning logs for canvas context errors

rhino-ty committed:
fix: resolve merge conflicts in App.tsx
